ThrusterController Class Review

ryan-summers opened the pull request.

gaofangfrank committed:
initial commit for thruster controller serial communication

gaofangfrank committed:
Moved thruster control files to movement

gaofangfrank committed:
functionality additions

gaofangfrank committed:
first working(?) commit

gaofangfrank committed:
Update and rename thruster_serial.cpp to maestro_class.cpp

gaofangfrank committed:
Rename thruster_serial.hpp to maestro_class.hpp

gaofangfrank committed:
explaination of class structure and functionality.

gaofangfrank committed:
minor changes

gaofangfrank committed:
added some (very) basic fix to circumvent ESC error issue

gaofangfrank committed:
added some documentation

gaofangfrank committed:
minor changes (i think)

gaofangfrank committed:
fixed normalized float value parsing

ryan-summers committed:
Addition of updated thruster class

ryan-summers committed:
Updating deconstructor to make use of vectors

ryan-summers committed:
Update of constructor to allow serial port to be set later.

ryan-summers committed:
Updating and correcting maestro controller

ryan-summers committed:
Removed an error!

ryan-summers committed:
Removed const from Speed

ryan-summers committed:
Correcting a bug in checking valid port

Commit:
Addition of reverse control capabilities

ryan-summers committed:
Addition of maestro class

skallaher commented on src/movement/maestro.hpp:
```diff
@@ -0,0 +1,342 @@
+         * Specifies the delay in seconds after which a reset signal (1500
+         * microsecond pulse) should be sent.
+         */
+        static constexpr double reset_delay = 120.0;
```
This could be made into a parameter on the parameter server.

ryan-summers replied:
It's never going to be a value that should be modified.

skallaher commented on src/movement/maestro.hpp:
```diff
@@ -0,0 +1,342 @@
+             *        and reverse directions that should be allowed by the
+             *        controller.
+             */
+            ThrusterController(const int _thrusters, const float _max_speed) :
```
Why is this not constructed with a port?

ryan-summers replied:
To allow for global variables before a serial port is defined. WE could also permit two constructors. IT doesn't need a valid serial object, it just needs a serial objeCT in general

skallaher commented on src/movement/maestro.hpp:
```diff
@@ -0,0 +1,342 @@
+         * microsecond pulse) should be sent.
+         */
+        static constexpr double reset_delay = 120.0;
+
```
Why is all of the code in the header? This class is not templated so there is no real reason for this.

ryan-summers replied:
It was originally templated, which is why all the code is here. Do we want to cut it into two files?

skallaher commented on src/movement/maestro.hpp:
```diff
@@ -0,0 +1,342 @@
+             *        controller.
+             */
+            ThrusterController(const int _thrusters, const float _max_speed) :
+                        port(NULL),
```
On lines 55, 69, 91, and 115 you use `NULL`. As this is in C++, and port is a pointer, you should be using `nullptr`.

ryan-summers replied:
Will update before pull

skallaher commented on src/movement/maestro.hpp:
```diff
@@ -0,0 +1,342 @@
+                if (port != NULL)
+                {
+                    vector<float> zero;
+                    for (int i = 0; i < thrusters; ++i)
```
See the documentation for a [C++ vector](http://www.cplusplus.com/reference/vector/vector/vector/). The fill constructor will perform this for you and fill a vector of size n with a default value.

skallaher commented on src/movement/maestro.hpp:
```diff
@@ -0,0 +1,342 @@
+             */
+            int setPort(Serial *_port)
+            {
+                port = _port;
```
C++ convention usually suggests that the underscore or m go before the member variable name, not the argument name.

ryan-summers replied:
I personally dislike that convention, but if it's in our coding atandards, I'll update it.

skallaher replied:
I would say it would be fine if the argument name did not have an underscore, however, for most this would cause confusion as they are backwards from convention.

skallaher commented on src/movement/maestro.hpp:
```diff
@@ -0,0 +1,342 @@
+                    uint8_t detect_byte = 0xAA;
+                    if (port->Write(&detect_byte, 1) != 1)
+                    {
+                        ROS_DEBUG_STREAM("Serial port failed to write baud-detection bit.");
```
This seems to be an error scope message. As does the message on line 101.

ryan-summers replied:
Ah okay, that makes sense. Will update messages in this file.

skallaher commented on src/movement/maestro.hpp:
```diff
@@ -0,0 +1,342 @@
+            {
+                if (speeds.size() != thrusters || port == NULL) return -1;
+
+                uint8_t command[thrusters*2+3];
```
Could you comment this line to explain what `thrusters * 2 + 3` is doing?

skallaher commented on src/movement/maestro.hpp:
```diff
@@ -0,0 +1,342 @@
+                if (now > nextReset)
+                {
+                    ROS_INFO_STREAM("Sending thruster reset signal.");
+                    for (int thruster = 0; thruster < thrusters; ++thruster)
```
These two variable names are too close to each other and can cause confusion.

skallaher commented on src/movement/maestro.hpp:
```diff
@@ -0,0 +1,342 @@
+                        if (parseNormalized(0, command[thruster*2+4],
+                                command[thruster*2+3]))
+                        {
+                            ROS_DEBUG_STREAM("Parse Normalized encountered
```
More messages that should probably be `ROS_ERROR`s. Lines 134, 143

skallaher commented on src/movement/maestro.hpp:
```diff
@@ -0,0 +1,342 @@
+                     * Experimental tests have found this value. Any value less
+                     * is insufficient and will cause ESC malfuctions.
+                     */
+                    usleep(185000);
```
Please make this value a parameter so that it is not hard coded.

ryan-summers replied:
Will add as parameter, since this is something that could potentially be updated later

skallaher commented on src/movement/maestro.hpp:
```diff
@@ -0,0 +1,342 @@
+                 */
+                for (int thruster = 0; thruster < speeds.size(); ++thruster)
+                {
+                    double speed = speeds[thruster];
```
More very similar names

skallaher commented on src/movement/maestro.hpp:
```diff
@@ -0,0 +1,342 @@
+                    if (parseNormalized(speed, command[2*thruster+4],
+                            command[2*thruster+3]))
+                    {
+                        ROS_DEBUG_STREAM("Parse Normalized encountered abnormal
```
Another `ROS_ERROR` candidate

skallaher commented on src/movement/maestro.hpp:
```diff
@@ -0,0 +1,342 @@
+                        sizeof(command));
+            }
+
+        private:
```
Declaring member variables at the top of a header file is easier to read. That way, the reader knows what the class contains without having to trudge through the rest of the file

skallaher commented on src/movement/maestro.hpp:
```diff
@@ -0,0 +1,342 @@
+             * Pointer to the serial port to be used for communication with the
+             * Maestro.
+             */
+            Serial *port;
```
These member variables are private. Convention suggests that their names begin with an underscore.

skallaher commented on src/movement/maestro.hpp:
```diff
@@ -0,0 +1,342 @@
+            /*
+             * The maximum speed setting that a thruster may be set to.
+             */
+            const float max_speed;
```
Please choose between underscore naming and camel case naming. Mixing the two is not advised.

ryan-summers committed:
Update of thruster controller class

ryan-summers committed:
Addition of ThrusterController changes post review

irwineffect commented on src/movement/ThrusterController.cpp:
```diff
@@ -0,0 +1,187 @@
+    ThrusterController::ThrusterController(const int thrusters,
+                                           const double max_speed,
+                                           Serial *port,
+                                           const double delay_ms) :
```
Constructor should be empty, pass in these parameters in an "init" function. This allows us to declare the class before we know how many thrusters there will be.

irwineffect commented on src/movement/ThrusterController.cpp:
```diff
@@ -0,0 +1,187 @@
+#include "movement/ThrusterController.hpp"
```
File should be renamed to include "maestro" in the name.

irwineffect commented on src/movement/ThrusterController.cpp:
```diff
@@ -0,0 +1,187 @@
+     * @param port The serial port to be used for communicating with the
+     *        thrusters. This port may be changed during runtime.
+     */
+    ThrusterController::ThrusterController(const int thrusters,
```
class should be renamed to include "maestro" in the name.

bkallaher commented on src/movement/ThrusterController.cpp:
```diff
@@ -0,0 +1,187 @@
+                                           Serial *port,
+                                           const double delay_ms) :
+            _port(port),
+            _thrusters(_thrusters),
```
Shouldn't this line say `_thrusters(thrusters),` not `_thrusters(_thrusters)`

irwineffect commented on src/movement/ThrusterController.cpp:
```diff
@@ -0,0 +1,187 @@
+     *
+     * @return 0 on success and -1 on failure.
+     */
+    int ThrusterController::set(vector<double> &speeds)
```
Simply taking in a vector of speeds will not work if we decide to disable a single thruster. Also, thrusters messages from up above are not necessarily in the order of ports. I recommend taking in a second vector of ports/channels that describe which motor the speed goes to.

irwineffect committed:
reworked maestro thruster controller class

irwineffect committed:
finished reworking thrusters

also cleaned up control system build, cobalt thruster settings, and upgraded serial_listener tool

bkallaher commented on src/movement/maestro_thruster_controller.cpp:
```diff
@@ -0,0 +1,207 @@
+            return -1;
+        }
+
+        if (speed > _max_speed[channel])
```
Use sgn to reduce these two statements into one

irwineffect replied:
done

bkallaher commented on src/movement/thruster_maestro.cpp:
```diff
@@ -1,51 +1,84 @@
+        result = thrusterController.set(msg->data[i], mThruster_info[i].channel);
+        if(result != 0)
+        {
+            ROS_ERROR_STREAM("Setting speed of thrusters failed.");
```
Add channel that failed into the error mesg

bkallaher commented on src/movement/thruster_maestro.cpp:
```diff
@@ -1,51 +1,84 @@
+        result = thrusterController.set(0, mThruster_info[i].channel);
+        if(result != 0)
+        {
+            ROS_ERROR_STREAM("Setting speed of thrusters failed.");
```
Add channel that failed into the error mesg

skallaher commented on src/movement/thruster_maestro.cpp:
```diff
@@ -1,51 +1,84 @@
+    if(!n.getParam("ports/thruster", thruster_port))
+    {
+        ROS_FATAL("no serial port specified, exiting!");
+        exit(1);
```
Shouldn't something like `ros:: shutdown()` be used here?

irwineffect replied:
As far as I can tell, ros::shutdown just sets the ros::ok flag. It doesn't directly cause the node to terminate. I was experiencing  hanging issues.

bkallaher commented on src/movement/thruster_maestro.cpp:
```diff
@@ -1,51 +1,84 @@
+        exit(1);
+    }
+
+    serial.Open(thruster_port.c_str(), B9600);
```
Maybe make the Baud rate a definition somewhere (param?). Not an issue but is it possible to run at a higher baud rate if necessary?

skallaher replied:
Possibly another `#define`?

skallaher commented on src/movement/thruster_maestro.cpp:
```diff
@@ -1,51 +1,84 @@
+    serial.Open(thruster_port.c_str(), B9600);
+
+    //Get the ports and names of the maestro thrusters (from coblat.yaml)
+    XmlRpc::XmlRpcValue my_list;
```
I feel like a better variable name could be used here instead of `my_list`

irwineffect replied:
yeah, good idea.

skallaher commented on src/movement/thruster_maestro.cpp:
```diff
@@ -1,51 +1,84 @@
+        ROS_DEBUG_STREAM("thrusters["<< i << "][name]:    " << my_list[i]["name"]);
+        ROS_DEBUG_STREAM("thrusters["<< i << "][channel]:    " << my_list[i]["channel"]);
+        Thruster_info one_thruster;
+        one_thruster.name = static_cast<std::string>(my_list[i]["name"]);
```
Would it be useful to add some type verification for these casts?

irwineffect replied:
What do you mean?

skallaher replied:
A static cast doesn't do any runtime checks to verify that this cast is valid. Though it's unlikely, someone could accidentally use the wrong type of value into the parameter causing problems with the cast.

skallaher commented on src/utility/serial.cpp:
```diff
@@ -28,7 +28,7 @@ namespace rs
         {
             ROS_FATAL("could not open port: %s", port_name);
-            ros::shutdown();
+            exit(1);
```
Is there a reason that `exit` is better than `ros::shutdown`?

irwineffect replied:
see above comment.

skallaher commented on tools/serial_listener.py:
```diff
@@ -12,46 +12,56 @@
-		
+
+    #setup clearscreen thread
+    import threading
```
I'm not very familiar with Python syntax and conventions, but shouldn't imports be done at the top of the file?

irwineffect replied:
sure, why not?

bkallaher replied:
In python if an import is done inside of a function its functionality is only available to the function. Python has scoped imports.

bkallaher commented on src/movement/thruster_maestro.cpp:
```diff
@@ -1,51 +1,84 @@
+    serial.Open(thruster_port.c_str(), B9600);
+
+    //Get the ports and names of the maestro thrusters (from coblat.yaml)
+    XmlRpc::XmlRpcValue my_list;
```
Can my_list have a more descriptive name?

bkallaher commented on src/movement/thruster_maestro.cpp:
```diff
@@ -1,51 +1,84 @@
-		outmessage.Open(thruster_port.c_str(), B9600);
-		ros::spin();
+
+    thrusterController.init(0.6, &serial);
```
What is 0.6? Magic number....

skallaher replied:
This is the max speed of the thrusters. @irwineffect didn't we talk about being able to set the max speeds of each thruster individually in the parameters?

bkallaher commented on src/movement/control_system.cpp:
```diff
@@ -1,5 +1,10 @@
 #include "control_system.hpp"
 
+int Sgn(float x)
```
Maybe move to a math library

irwineffect replied:
done

bkallaher commented on src/movement/maestro_thruster_controller.cpp:
```diff
@@ -0,0 +1,207 @@
+            _max_speed[i] = max_speed;
+        }
+
+        if (delay_ms < 185.00)
```
185.0 is a magic number ...

skallaher replied:
Apparently this was found empirically, though I agree it probably should be `#defined` or made a constant.

irwineffect replied:
done

bkallaher commented on src/movement/maestro_thruster_controller.cpp:
```diff
@@ -0,0 +1,207 @@
+     *
+     * @return Zero on success and -1 on failure.
+     */
+     int MaestroThrusterController::parseNormalized(const double speed, uint8_t &msb,
```
400, 1500, and 4 seem like magic numbers...

irwineffect replied:
I'm keeping these in here for clarity, but I'll add comments to clarify what they mean.

bkallaher commented on src/movement/maestro_thruster_controller.hpp:
```diff
@@ -0,0 +1,186 @@
+        ~MaestroThrusterController();
+
+        int init(const double max_speed, Serial *port,
+            const double post_reset_delay_ms = 185.00);
```
185.0 again...

irwineffect committed:
implemented changes from skallaher

irwineffect committed:
implemented fixes requested by brandon

skallaher commented on src/movement/maestro_thruster_controller.cpp:
```diff
@@ -56,7 +57,8 @@ namespace rs
         }
 
-        if (delay_ms < 185.00)
+        _post_reset_delay_ms == post_reset_delay_ms;
```
Is this meant to be an assignment? It is currently a comparison.

irwineffect replied:
fixed in 0ac45a2

irwineffect committed:
fixed assignment bug
